src/python/pants/init/extension_loader.py

```diff
@@ -28,12 +31,23 @@ class PluginLoadOrderError(PluginLoadingError):
     pass
 
 
-def load_backends_and_plugins(
-    plugins: List[str],
-    working_set: WorkingSet,
-    backends: List[str],
-    bc_builder: Optional[BuildConfiguration.Builder] = None,
-) -> BuildConfiguration:
+@dataclass(frozen=True)
+class TemplatedBackendConfig:
+    template: str
+    kwargs: FrozenDict[str, Any]
+
+    @classmethod
+    def from_dict(cls, d: Any):
+        d = dict(d)
+        template = d.pop('template', None)
+        if not template:
+            raise ValueError(f'"template" is a required key for a backend template')
+        return cls(template=cast(str, template), kwargs=FrozenDict(d))
```

Comment:
I think it would make sense to move this to `global_options.py` and have a method there that returns the templated backends mapping, rather than building it in options_initializer.

Reply:
I've been digging for a way to do this and my first thought was:
```python
class TemplateBackendsOption(DictOption[TemplatedBackendConfig]):
    def _convert_(self, val: Any) -> dict[str, TemplatedBackendConfig]:
        values = {backend_alias: TemplatedBackendConfig.from_dict(templating_config)
                  for backend_alias, templating_config in val.items()}
        return super()._convert_(values)
```
and then setting 
```python
templated_backends = TemplateBackendsOption(
    advanced=True,
    default={},
    help="TODO",
)
```

but I am finding that this `_convert_` method does not seem to be called when preparing the value that gets returned by `bootstrap_options.templated_backends` where:
```python
options_bootstrapper: OptionsBootstrapper
bootstrap_options = options_bootstrapper.get_bootstrap_options().for_global_scope()
bootstrap_options.templated_backends
```
Instead I end up still with the raw dict on the key side of the Option. Debugging through `pants.option.parser` I am finding that `Parser._compute_value` does not seem to make use of the whole Option descriptor but just the kwargs `{'help': 'TODO', 'default': {}, 'type': <class 'dict'>, 'advanced': True}`

@kaos Do you know of examples of `ListOption` or `DictOption` that convert their member (or value) types into something aside from what came through from the literal parse of the TOML?

Reply:
yes, sorry for not providing an example up front. I imagined something like this: 
https://github.com/pantsbuild/pants/blob/45e7be7c81916dfcf23fb2233ce760949c2c8249/src/python/pants/backend/docker/subsystems/docker_options.py#L258-L260

Reply:
Interesting. I wonder if that kind of functionality is available during the bootstrapping phase. During backend loading we don't pass around Subsystem objects (which are available in the engine via the [construct subsystem rules](https://github.com/pantsbuild/pants/blob/a6c55c149918de2893a973225b10b7b1c4d60294/src/python/pants/option/subsystem.py#L311)) but rather plain `OptionValueContainer`. I haven't figured out what to hang a method or property off of that would be accessible within, say https://github.com/pantsbuild/pants/blob/fd02447a750b8629826d5217146bdebd312c0470/src/python/pants/init/options_initializer.py#L38

Reply:
The `bootstrap_options` here:
https://github.com/pantsbuild/pants/blob/fd02447a750b8629826d5217146bdebd312c0470/src/python/pants/init/options_initializer.py#L50

is an instance of:
https://github.com/pantsbuild/pants/blob/fd02447a750b8629826d5217146bdebd312c0470/src/python/pants/option/global_options.py#L913

Which is where this ought to live ;)

Reply:
I think...

Reply:
Oh I see... so it's not actually passing an instance of the BootstrapOptions or GlobalOptions, but an Options container with the various options from the class registered... huh, weird.

Reply:
OK, so how about a compromise, something along the lines of a classmethod `GlobalOptions.get_backend_template_configs(bootstrap_options.templated_backends)` ?

Reply:
Made the change. I think it looks nice this way you suggested. https://github.com/pantsbuild/pants/pull/20270/commits/3d9aa7555e4e186231ed03553adf78b97a23a8d4

Reply:
Resolving for now, but we can reopen/revisit if needed